### ISSUE TITLE
Autoscaling group tags are destroyed and recreated for an update in-place

### ISSUE THREAD

**Issue by Lynty** [hashicorp/terraform-provider-aws#15550]
Relates to: [#25949](https://github.com/hashicorp/terraform/issues/25949)
This originally was an issue using `provider.aws v2.52.0`. It was thought to be fixed with `provider.aws v3.5.0` but still coming across the same issue. 

### Terraform Version

Hello,
I have a Service Control Policy configured in an AWS account that denies creation and updating of an autoscaling group if it does not have some mandatory tags. I am using this AWS [example](https://docs.aws.amazon.com/organizations/latest/userguide/orgs_manage_policies_example-scps.html#example-require-tag-on-create) to accomplish this. It uses a Null condition operator to check if a condition key is present at the time of authorization. If it is `true` (the key doesn't exist — it is null), the SCP prevents resource creation.

The issue I am coming across is that regardless of a valid or invalid change to a tag, a `terraform plan` output shows a change in place for the autoscaling group resource but actually destroys and recreates the changed tag. A `terraform apply` will trigger the SCP and deny updating the autoscaling group due to the tag being momentarily destroyed/marked as `null`.

```
Terraform v0.12.24
+ provider.aws v3.5.0
```

### Terraform Configuration Files

```terraform
resource "aws_autoscaling_group" "webserver" {
  name_prefix               = "webserver"
  max_size                  = 1
  min_size                  = 0
  desired_capacity          = 1
  health_check_grace_period = 300
  health_check_type         = "EC2"
  force_delete              = true

  lifecycle {
    create_before_destroy = true
  }

  launch_configuration = aws_launch_configuration.webserver.name

  vpc_zone_identifier = var.public_subnets

  tag {
    key                 = "environment"
    value               = var.env
    propagate_at_launch = "true"
  }
}

```
Service Control Policy Document
```
data "aws_iam_policy_document" "scp_mandatory_ec2_tags" {
  version = "2012-10-17"
  statement {
    sid = "DenyAutoscalingGroupWithNoEnvironmentTag"

    actions = [
      "autoscaling:CreateAutoScalingGroup",
      "autoscaling:UpdateAutoScalingGroup",
    ]

    effect = "Deny"

    resources = [
      "arn:aws:autoscaling:*:*:autoScalingGroup:*:autoScalingGroupName/*",
    ]

    condition {
      test     = "Null"
      variable = "aws:RequestTag/environment"

      values = [
        "true",
      ]
    }
  }
}
```

### Expected Behavior

`environment` is a mandatory tag. I should spin up an ec2 instance in an autoscaling group with tag `environment:dev`. While the instance is running, I should be able to change the tag value to something like `environment:stage` without triggering the Service Control Policy with terraform.

### Actual Behavior

SCP is triggered and blocks autoscaling resource update due to seeing a tag being destroyed/marked as `null`. 
`terraform plan`
```
Refreshing Terraform state in-memory prior to plan...
The refreshed state will be used to calculate this plan, but will not be
persisted to local or remote state storage.

aws_autoscaling_group.webserver: Refreshing state... [id=webserver20200820173423399100000001]

------------------------------------------------------------------------

An execution plan has been generated and is shown below.
Resource actions are indicated with the following symbols:
  ~ update in-place

Terraform will perform the following actions:

  # aws_autoscaling_group.webserver will be updated in-place
  ~ resource "aws_autoscaling_group" "webserver" {
        arn                       = "arn:aws:autoscaling:us-west-2:xxxx:autoScalingGroup:624943a4-2ab9-4585-90ca-c30a185c14d3:autoScalingGroupName/webserver20200820173423399100000001"
        availability_zones        = [
            "us-west-2a",
            "us-west-2b",
        ]
        default_cooldown          = 300
        desired_capacity          = 1
        enabled_metrics           = []
        force_delete              = true
        health_check_grace_period = 300
        health_check_type         = "EC2"
        id                        = "webserver20200820173423399100000001"
        launch_configuration      = "webserver20200819192444336200000002"
        load_balancers            = []
        max_instance_lifetime     = 0
        max_size                  = 1
        metrics_granularity       = "1Minute"
        min_size                  = 0
        name                      = "webserver20200820173423399100000001"
        name_prefix               = "webserver"
        protect_from_scale_in     = false
        service_linked_role_arn   = "arn:aws:iam::xxxx:role/aws-service-role/autoscaling.amazonaws.com/AWSServiceRoleForAutoScaling"
        suspended_processes       = []
        target_group_arns         = []
        termination_policies      = []
        vpc_zone_identifier       = [
            "subnet-xxxx",
            "subnet-xxxx",
        ]
        wait_for_capacity_timeout = "10m"

      - tag {
          - key                 = "environment" -> null
          - propagate_at_launch = true -> null
          - value               = "dev" -> null
        }
      + tag {
          + key                 = "environment"
          + propagate_at_launch = true
          + value               = "stage"
        }
    }

Plan: 0 to add, 1 to change, 0 to destroy.

------------------------------------------------------------------------

Note: You didn't specify an "-out" parameter to save this plan, so Terraform
can't guarantee that exactly these actions will be performed if
"terraform apply" is subsequently run.
```
```
Error: Error updating Autoscaling group: AccessDenied: User: arn:aws:sts::xxxx:assumed-role/OrganizationAccountAccessRole/xxxx is not authorized to perform: autoscaling:UpdateAutoScalingGroup on resource: arn:aws:autoscaling:us-west-2:xxxx:autoS
calingGroup:bec94b18-e890-4d68-ad68-b23347793ea6:autoScalingGroupName/webserver20201007192319493500000001 with an explicit deny status code: 403, request id: cf05d073-e712-4b4d-8e60-ad763ffe3561
```
### Steps to Reproduce

1. `terraform init`
2. `terraform apply`
3. Change a tag value to a different valid value
4. `terraform apply`

**Comment by anGie44** [comment]
Hi @Lynty, thank you for writing up this issue! Looking at the error log, i'm interested to know if after running the `apply` and receiving the error, did the tags on the autoscaling group still get updated? I ask this because looking at the provider code, when it comes to updating an `autoscaling_group` resource, the `tags` on the resource should actually get updated first via a call to `CreateOrUpdateTags` and then additional changes are made with a call to `UpdateAutoScalingGroup`. Which then also leads me to believe that perhaps the `aws_iam_policy_document` data-source should use `CreateOrUpdateTags` in the `actions` list instead of `UpdateAutoScalingGroup` since this latter method doesn't have a `Tags` field in the request params (API reference: https://docs.aws.amazon.com/autoscaling/ec2/APIReference/API_UpdateAutoScalingGroup.html) and therefore will cause an Update to consistently fail.

**Comment by maishiroma** [comment]
Hello @anGie44! 

Thank you for your response! I apologize for the long wait on the reply from Lynty. She is no longer involved in the current project she saw this behavior in and combined with other project priorities, our team have not had the change to revisit this issue until recently. Moving forward, I will be the point of contract for this issue instead. 

Anyway, as to your suggestion, I went ahead and updated our Service Control Principle to use `autoscale:CreateOrUpdateTags` to enforce our expected behavior: 

```
data "aws_iam_policy_document" "scp_mandatory_ec2_tags" {
  version = "2012-10-17"
  statement {
    sid = "DenyAutoscalingGroupWithNoEnvironmentTag"

    actions = [
      "autoscaling:CreateAutoScalingGroup",
      "autoscaling:CreateOrUpdateTags",
    ]

    effect = "Deny"

    resources = [
      "arn:aws:autoscaling:*:*:autoScalingGroup:*:autoScalingGroupName/*",
    ]

    condition {
      test     = "Null"
      variable = "aws:RequestTag/environment"

      values = [
        "true",
      ]
    }
  }
}
```

After I applied this change, I ran a `terraform and apply` using AWS provider `3.5.0`, to which I am still seeing this same behavior:

```
$ terraform apply

An execution plan has been generated and is shown below.
Resource actions are indicated with the following symbols:
  ~ update in-place

Terraform will perform the following actions:

  # aws_autoscaling_group.webserver will be updated in-place
  ~ resource "aws_autoscaling_group" "webserver" {
        arn                       = "arn:aws:autoscaling:us-west-2:xxxxxxx:autoScalingGroup:a12f3014-2ee9-4227-bc1a-2be63a2a5ae4:autoScalingGroupName/webserver2020111116095628750000003a"
        availability_zones        = [
            "us-west-2a",
            "us-west-2b",
        ]
        default_cooldown          = 300
        desired_capacity          = 1
        enabled_metrics           = []
        force_delete              = true
        health_check_grace_period = 300
        health_check_type         = "EC2"
        id                        = "webserver2020111116095628750000003a"
        launch_configuration      = "webserver20201111160946391600000035"
        load_balancers            = []
        max_instance_lifetime     = 0
        max_size                  = 1
        metrics_granularity       = "1Minute"
        min_size                  = 0
        name                      = "webserver2020111116095628750000003a"
        name_prefix               = "webserver"
        protect_from_scale_in     = false
        service_linked_role_arn   = "arn:aws:iam::xxxxxxx:role/aws-service-role/autoscaling.amazonaws.com/AWSServiceRoleForAutoScaling"
        suspended_processes       = []
        target_group_arns         = []
        termination_policies      = []
        vpc_zone_identifier       = [
            "subnet-xxxxxx",
            "subnet-xxxxxx",
        ]
        wait_for_capacity_timeout = "10m"

      - tag {
          - key                 = "environment" -> null
          - propagate_at_launch = true -> null
          - value               = "opsdev" -> null
        }
      + tag {
          + key                 = "environment"
          + propagate_at_launch = true
          + value               = "tst"
        }
    }

Plan: 0 to add, 1 to change, 0 to destroy.

Do you want to perform these actions?
  Terraform will perform the actions described above.
  Only 'yes' will be accepted to approve.

  Enter a value: yes

aws_autoscaling_group.webserver: Modifying... [id=webserver2020111116095628750000003a]

Error: error updating tags for Auto Scaling Group (webserver2020111116095628750000003a): error tagging resource (webserver2020111116095628750000003a): AccessDenied: User: arn:aws:sts::xxxxxxx:assumed-role/OrganizationAccountAccessRole/xxxxxxxxx is not authorized to perform: autoscaling:CreateOrUpdateTags on resource: arn:aws:autoscaling:us-west-2:xxxxxxxx:autoScalingGroup:a12f3014-2ee9-4227-bc1a-2be63a2a5ae4:autoScalingGroupName/webserver2020111116095628750000003a with an explicit deny
        status code: 403, request id: 2151f9a6-e8b5-4ce5-97fc-1775f3134345
```
I reinitialized my directory and removed the `.terraform` folder and I am still seeing this behavior.

In other words, it still seems that the behavior that we were seeing is still present in 3.5.0.

**Comment by maishiroma** [comment]
Upon looking at this problem a bit more, I think the main reason that this is not working is because the ASG resource is not using that native AWS API call to [update or release tags](https://docs.aws.amazon.com/autoscaling/ec2/APIReference/API_CreateOrUpdateTags.html).

If one looks in the function update for updating ASGs for both [v3.5.0](https://github.com/hashicorp/terraform-provider-aws/blob/v3.5.0/aws/resource_aws_autoscaling_group.go#L986) (which is also the same in the current [release](https://github.com/hashicorp/terraform-provider-aws/blob/master/aws/resource_aws_autoscaling_group.go#L1006)), it is doing a map merge. 
```
if d.HasChanges("tag", "tags") {
		oTagRaw, nTagRaw := d.GetChange("tag")
		oTagsRaw, nTagsRaw := d.GetChange("tags")

		oTag := keyvaluetags.AutoscalingKeyValueTags(oTagRaw, d.Id(), autoscalingTagResourceTypeAutoScalingGroup)
		oTags := keyvaluetags.AutoscalingKeyValueTags(oTagsRaw, d.Id(), autoscalingTagResourceTypeAutoScalingGroup)
		oldTags := oTag.Merge(oTags).AutoscalingTags()

		nTag := keyvaluetags.AutoscalingKeyValueTags(nTagRaw, d.Id(), autoscalingTagResourceTypeAutoScalingGroup)
		nTags := keyvaluetags.AutoscalingKeyValueTags(nTagsRaw, d.Id(), autoscalingTagResourceTypeAutoScalingGroup)
		newTags := nTag.Merge(nTags).AutoscalingTags()

		if err := keyvaluetags.AutoscalingUpdateTags(conn, d.Id(), autoscalingTagResourceTypeAutoScalingGroup, oldTags, newTags); err != nil {
			return fmt.Errorf("error updating tags for Auto Scaling Group (%s): %w", d.Id(), err)
		}
	}
```
In other words, it's using custom code to get the difference between the planned and the currently applied tags, which makes sense on why a tag isn't being replaced in line. 

As this [comment](https://github.com/hashicorp/terraform-provider-aws/issues/14768#issuecomment-678345040) originally brought up the next steps if the new logic doesn't solve this issue, so @ewbankkit and @bflad, what would be the path from here?
